### PR TITLE
[SPMD] Remove the code branches only used in SPMD mode from distributed.py

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55353 [SPMD] Remove the code branches only used in SPMD mode from distributed.py**

Remove all the code branches that will only be executed when `device_ids > 1`.

Some helper functions are also removed:
1.  `_verify_replicas_within_process` and `verify_replicas_within_process`
2. `_replicate_modules_within_process`
3. `parallel_apply`

The next step is deprecating `_module_copies` field.

Differential Revision: [D27552201](https://our.internmc.facebook.com/intern/diff/D27552201/)

### CODE REVIEW

**Review by pritamdamania87 — COMMENTED** [review]
Nice work removing all the codepaths and making DDP code much simpler :) Can you double check the CI failures though?

**Comment by SciPioneer** [comment]
> Nice work removing all the codepaths and making DDP code much simpler :) Can you double check the CI failures though?

Last time I double checked the failures, and did not find any relevant ones. Rebased and run the tests again.

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -480,16 +476,12 @@ def __init__(
...
         )
 
+        # TODO(wayi@): Remove this field since SPMD is no longer supported,
+        # and also remove all the relevant unnecessary loops.
         # Module replication within process (single-process multi device)
-        self._module_copies = self._replicate_modules_within_process()
+        self._module_copies = [self.module]
         # Build parameters for reducer.
         parameters, expect_sparse_gradient = self._build_params_for_reducer()
-        # Verify model equivalence.
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:487]: Keep the relevant part of these comments (we still need to verify model equivalence across processes)?

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -570,91 +562,26 @@ def __setstate__(self, state):
...
-                        # Note that we access module.named_parameters instead of
-                        # parameters(module). parameters(module) is only needed in the
-                        # single-process multi device case, where it accesses replicated
-                        # parameters through _former_parameters.
-                        for param_name, param in module.named_parameters(recurse=False)
-                        if param.requires_grad
-                        and f"{module_name}.{param_name}"
-                        not in self.parameters_to_ignore
-                    ]
+        modules_and_parameters = [
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:567]: This outermost list is always one element now? I guess to flatten in we'd need to change the reducer signature though?

File: torch/nn/parallel/distributed.py [hunk 3]
```diff
@@ -1330,9 +1204,7 @@ def _passing_sync_batchnorm_handle(self, module_copies):
                     assert (
                         self.device_type != "cpu"
                     ), "SyncBatchNorm layers only work with GPU modules"
-                    layer._specify_ddp_gpu_num(
-                        len(self.device_ids) if self.device_ids else 1
-                    )
+                    layer._specify_ddp_gpu_num(1)
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:1207]: I wonder if we can always hardcode this to 1 now in sync batchnorm. Looking at the SyncBN implementation, IIUC it looks like setting this attribute is what lets the layer know we are training in DDP (which isn't very robust either since anyone can set it it seems like). So we would need to figure out some way to ensure that SyncBN is only used with DDP if we decide that we would like to remove these calls.

**Review by rohan-varma — APPROVED** [review]
LGTM, tests look clear. Please see a couple comments

File: torch/nn/parallel/distributed.py (hunk 2 above)
Comment (SciPioneer) [reply torch/nn/parallel/distributed.py:567]: As the description of this PR says, I plan to deprecate` _module_copies` field, which is the root cause of the extra complexity here. Will simplify it in the next PR.

File: torch/nn/parallel/distributed.py (hunk 3 above)
Comment (SciPioneer) [reply torch/nn/parallel/distributed.py:1208]: Thanks for the suggestion! Plan to remove `_specify_ddp_gpu_num` method later, as it is also used by fairscale.

File: torch/nn/parallel/distributed.py [hunk 4]
```diff
@@ -568,93 +560,29 @@ def __setstate__(self, state):
         self.__dict__.setdefault("require_forward_param_sync", True)
         self.__dict__.setdefault("require_backward_grad_sync", True)
         parameters, expect_sparse_gradient = self._build_params_for_reducer()
+        # Verify model equivalence.
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:563]: I think this doc should go before `dist._verify_model_across_ranks` a few lines above

Comment (SciPioneer) [reply torch/nn/parallel/distributed.py:563]: Thanks for catching this! Didn't realize that how this comment went to that line.

Corrected in #55877

File: torch/nn/parallel/distributed.py (hunk 3 above)
Comment (SciPioneer) [reply torch/nn/parallel/distributed.py:1208]: Resolved by #56425 and #55946.
